attempt to streamline format hook

mmacpherson opened the pull request.

mmacpherson committed:
attempt to streamline format hook

scrambldchannel commented:
Hi @mmacpherson, thanks for this! I haven't looked at Julia or this hook for a long time.... :) I might need a bit of time to get my head around this. It might help if you give me a quick overview of what it is doing?

mmacpherson committed:
annotate code

mmacpherson commented:
Thanks for the comment @scrambldchannel.

Please excuse; I've now added some comments that outline what the code is doing at the line level.

I had been trying to speed up the hook. I used the fact that pre-commit passes in a list of staged julia files, to see if it mattered whether we only processed changed files, as opposed to all of them. And I added a check around the JuliaFormatter import, to see if reinstalling the package on each invocation was slowing things down.

This does seem to help somle, but it's still fairly slow. It seems to be one of these cases where one would need to compile a custom Julia image with JuliaFormatter precompiled. Or perhaps have some long-running server. At any rate stuff that seems out of scope for a pre-commit hook.

Thanks for putting this hook out there!

scrambldchannel commented:
Thanks for the feedback @mmacpherson, let's just merge it!

And yes, it is slow, that's just part of the trade-offs Julia makes really and isn't really suited for a pre-commit hook. If someone were to write a Julia formatter in e.g. Python, then it would probably be more feasible.

Anyway, I'm glad it was of use to you. Let's just get it merged, what's the worst that could happen? ;)
